File: CPP/astar_class.cpp
#include "includes/astar_class.h"
#include <iostream>

// ---------------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------------
Astar::Astar(Node start, std::pair<double, double> goal, int clearance, int radius)
    : start(start), goal(goal), clearance(clearance), radius(radius),
      numRows(200), numCols(300), move_i(0.0), move_j(0.0),
      goalRow(0.0), goalCol(0.0)
{
}

// ---------------------------------------------------------------------------
// IsValid
// ---------------------------------------------------------------------------
bool Astar::IsValid(double currRow, double currCol)
{
    double rowMin = 1.0 + radius + clearance;
    double rowMax = numRows - radius - clearance;
    double colMin = 1.0 + radius + clearance;
    double colMax = numCols - radius - clearance;
    return currRow >= rowMin && currRow <= rowMax && currCol >= colMin && currCol <= colMax;
}

// ---------------------------------------------------------------------------
// IsObstacle
// ---------------------------------------------------------------------------
bool Astar::IsObstacle(double row, double col)
{
    double sum_of_c_and_r = clearance + radius;
    double sqrt_of_c_and_r = 1.4142 * sum_of_c_and_r;

    // Get the distance from the starting position to the edge of the circle obstacle, subtracting (clearance + radius)
    double dist1 = sqrt((row - circle_row) * (row - circle_row) + (col - circle_col) * (col - circle_col)) - (circle_radius + sum_of_c_and_r);

    // Ellipse: inflate semi-axes by sum_of_c_and_r (matches Python)
    double dx = col - ellipse_col;
    double dy = row - ellipse_row;
    double a_inf = ellipse_a + sum_of_c_and_r;
    double b_inf = ellipse_b + sum_of_c_and_r;
    double dist2 = (dy * dy) / (b_inf * b_inf) + (dx * dx) / (a_inf * a_inf) - 1.0;

    // Triangle 1: vertices inflated to match Python
    double t1_x1 = 120.0 - 2.62 * sum_of_c_and_r, t1_y1 = 20.0 - 1.205 * sum_of_c_and_r;
    double t1_x2 = 150.0 - sqrt_of_c_and_r,        t1_y2 = 50.0;
    double t1_x3 = 185.0 + sum_of_c_and_r,          t1_y3 = 25.0 - sum_of_c_and_r * 0.9247;
    double t1_e1 = ((col - t1_y1) * (t1_x2 - t1_x1)) - ((t1_y2 - t1_y1) * (row - t1_x1));
    double t1_e2 = ((col - t1_y2) * (t1_x3 - t1_x2)) - ((t1_y3 - t1_y2) * (row - t1_x2));
    double t1_e3 = ((col - t1_y3) * (t1_x1 - t1_x3)) - ((t1_y1 - t1_y3) * (row - t1_x3));
    double dist3 = 1.0;
    if (t1_e1 <= 0.0 && t1_e2 <= 0.0 && t1_e3 <= 0.0) dist3 = 0.0;

    // Triangle 2: vertices inflated to match Python
    double t2_x1 = 150.0 - sqrt_of_c_and_r,  t2_y1 = 50.0;
    double t2_x2 = 185.0 + sum_of_c_and_r,    t2_y2 = 25.0 - sum_of_c_and_r * 0.9247;
    double t2_x3 = 185.0 + sum_of_c_and_r,    t2_y3 = 75.0 + sum_of_c_and_r * 0.5148;
    double t2_e1 = ((col - t2_y1) * (t2_x2 - t2_x1)) - ((t2_y2 - t2_y1) * (row - t2_x1));
    double t2_e2 = ((col - t2_y2) * (t2_x3 - t2_x2)) - ((t2_y3 - t2_y2) * (row - t2_x2));
    double t2_e3 = ((col - t2_y3) * (t2_x1 - t2_x3)) - ((t2_y1 - t2_y3) * (row - t2_x3));
    double dist4 = 1.0;
    if (t2_e1 >= 0.0 && t2_e2 >= 0.0 && t2_e3 >= 0.0) dist4 = 0.0;

    // Rhombus: convex obstacle collision check (diagonal-aligned)
    // check rhombus (half-plane test; dist5/dist6 = 0 if inside else 1)
    double r_x1 = (10.0 - sqrt_of_c_and_r);
    double r_y1 = 225.0;
    double r_x2 = 25.0;
    double r_y2 = (200.0 - sqrt_of_c_and_r);
    double r_x3 = (40.0 + sqrt_of_c_and_r); 
    double r_y3 = 225.0;
    double r_x4 = 25.0; 
    double r_y4 = (250.0 + sqrt_of_c_and_r);

    double first = ((col - r_y1) * (r_x2 - r_x1)) - ((r_y2 - r_y1) * (row - r_x1));
    double second = ((col - r_y2) * (r_x3 - r_x2)) - ((r_y3 - r_y2) * (row - r_x2));
    double third = ((col - r_y3) * (r_x4 - r_x3)) - ((r_y4 - r_y3) * (row - r_x3));
    double fourth = ((col - r_y4) * (r_x1 - r_x4)) - ((r_y1 - r_y4) * (row - r_x4));

    double dist5 = 1.0;
    double dist6 = 1.0;

    if (first >= 0.0 && second >= 0.0 && third >= 0.0 && fourth >= 0.0)
    {
        dist5 = 0.0;
        dist6 = 0.0;
    }

    // Square: collision check using axis-aligned bounding constraints (AABB)
    double s_x1 = (150.0 - sqrt_of_c_and_r), s_y1 = 50.0;
    double s_x2 = (120.0 - sqrt_of_c_and_r), s_y2 = 75.0;
    double s_x3 = 150.0, s_y3 = (100.0 + sqrt_of_c_and_r);
    double s_x4 = (185.0 + sum_of_c_and_r), s_y4 = (75.0 + (sum_of_c_and_r * 0.5148));

    double sq_first  = ((col - s_y1) * (s_x2 - s_x1)) - ((s_y2 - s_y1) * (row - s_x1));
    double sq_second = ((col - s_y2) * (s_x3 - s_x2)) - ((s_y3 - s_y2) * (row - s_x2));
    double sq_third  = ((col - s_y3) * (s_x4 - s_x3)) - ((s_y4 - s_y3) * (row - s_x3));
    double sq_fourth = ((col - s_y4) * (s_x1 - s_x4)) - ((s_y1 - s_y4) * (row - s_x4));

    double dist7 = 1.0;
    double dist8 = 1.0;

    if (sq_first <= 0.0 && sq_second <= 0.0 && sq_third <= 0.0 && sq_fourth <= 0.0)
    {
        dist7 = 0.0;
        dist8 = 0.0;
    }

    // Rod: collision check for elongated rectangular obstacle (AABB)
    double rod_first  = ((col - 95.0) * (8.66 + sqrt_of_c_and_r)) - ((5.0 + sqrt_of_c_and_r) * (row - 30.0 + sqrt_of_c_and_r));
    double rod_second = ((col - 95.0) * (37.5 + sqrt_of_c_and_r)) - ((-64.95 - sqrt_of_c_and_r) * (row - 30.0 + sqrt_of_c_and_r));
    double rod_third  = ((col - 30.05 + sqrt_of_c_and_r) * (8.65 + sqrt_of_c_and_r)) - ((5.45 + sqrt_of_c_and_r) * (row - 67.5));
    double rod_fourth = ((col - 35.5) * (-37.49 - sqrt_of_c_and_r)) - ((64.5 + sqrt_of_c_and_r) * (row - 76.15 - sqrt_of_c_and_r));

    double dist9 = 1.0;
    double dist10 = 1.0;

    if (rod_first <= 0.0 && rod_second >= 0.0 && rod_third >= 0.0 && rod_fourth >= 0.0)
    {
        dist9 = 0.0;
        dist10 = 0.0;
    }

    if (dist1 <= 0.0 || dist2 <= 0.0 ||
        dist3 == 0.0 || dist4 == 0.0 ||
        dist5 == 0.0 || dist6 == 0.0 ||
        dist7 == 0.0 || dist8 == 0.0 ||
        dist9 == 0.0 || dist10 == 0.0)
    {
        return true;
    }
    return false;
}

// ---------------------------------------------------------------------------
// Action methods
// Each method:
//   1. Computes the new heading angle
//   2. Sets move_i = sin(rad), move_j = cos(rad)
//   3. Returns IsValid(nextRow, nextCol) && !IsObstacle(nextRow, nextCol)
// ---------------------------------------------------------------------------
bool Astar::ActionMove30Up(double currRow, double currCol, int currAngle)
{
    int theta = (currAngle + 30) % 360;
    double rad = theta * M_PI / 180.0;
    move_i = std::sin(rad);
    move_j = std::cos(rad);
    return IsValid(currRow + move_i, currCol + move_j) && !IsObstacle(currRow + move_i, currCol + move_j);
}

bool Astar::ActionMove60Up(double currRow, double currCol, int currAngle)
{
    int theta = (currAngle + 60) % 360;
    double rad = theta * M_PI / 180.0;
    move_i = std::sin(rad);
    move_j = std::cos(rad);
    return IsValid(currRow + move_i, currCol + move_j) && !IsObstacle(currRow + move_i, currCol + move_j);
}

bool Astar::ActionMove30Down(double currRow, double currCol, int currAngle)
{
    int theta = (currAngle - 30) % 360;
    double rad = theta * M_PI / 180.0;
    move_i = std::sin(rad);
    move_j = std::cos(rad);
    return IsValid(currRow + move_i, currCol + move_j) && !IsObstacle(currRow + move_i, currCol + move_j);
}

bool Astar::ActionMove60Down(double currRow, double currCol, int currAngle)
{
    int theta = (currAngle - 60) % 360;
    double rad = theta * M_PI / 180.0;
    move_i = std::sin(rad);
    move_j = std::cos(rad);
    return IsValid(currRow + move_i, currCol + move_j) && !IsObstacle(currRow + move_i, currCol + move_j);
}

bool Astar::ActionMoveStraight(double currRow, double currCol, int currAngle)
{
    int theta = currAngle;
    double rad = theta * M_PI / 180.0;
    move_i = std::sin(rad);
    move_j = std::cos(rad);
    return IsValid(currRow + move_i, currCol + move_j) && !IsObstacle(currRow + move_i, currCol + move_j);
}

// ---------------------------------------------------------------------------
// CheckIfGoal
// ---------------------------------------------------------------------------
bool Astar::CheckIfGoal(double currRow, double currCol)
{
    double check = (currRow - goal.first) * (currRow - goal.first) + (currCol - goal.second) * (currCol - goal.second) - 1.5 * 1.5;
    if (check <= 0)
    {
        goalRow = currRow;
        goalCol = currCol;
        std::cout << "goal reached\n";
        return true;
    }
    return false;
}

// ---------------------------------------------------------------------------
// Run  (A* search)
// ---------------------------------------------------------------------------
std::tuple<std::vector<Node>, std::vector<Node>, double> Astar::Run()
{
    std::unordered_map<Node, double, NodeHash> distMap;
    std::unordered_map<Node, Node, NodeHash> pathMap;
    std::unordered_map<Node, bool, NodeHash> visited;

    for (double row = 1.0; row <= numRows; row += 0.5)
    {
        for (double col = 1.0; col <= numCols; col += 0.5)
        {
            for (int angle = 0; angle < 360; angle += 30)
            {
                Node n = {row, col, angle};
                distMap[n] = std::numeric_limits<double>::infinity();
                pathMap[n] = {-1.0, -1.0, -1}; // SENTINEL
                visited[n] = false;
            }
        }
    }

    std::vector<Node> explored_states;
    std::priority_queue<PQEntry, std::vector<PQEntry>, std::greater<PQEntry>> pq;

    distMap[start] = 0.0;
    pq.push({0.0, start});

    while (!pq.empty())
    {
        auto [f_cost, currNode] = pq.top();
        pq.pop();

        if (visited[currNode])
            continue;
        visited[currNode] = true;
        explored_states.push_back(currNode);

        auto [currRow, currCol, currAngle] = currNode;

        if (CheckIfGoal(currRow, currCol))
            break;

        int action1 = (currAngle + 30) % 360;
        int action2 = (currAngle + 60) % 360;
        int action3 = ((currAngle - 30) % 360 + 360) % 360;
        int action4 = ((currAngle - 60) % 360 + 360) % 360;

        for (const auto &[action_valid, next_angle, step_cost] : std::vector<std::tuple<bool, int, double>>{
                 {ActionMove30Up(currRow, currCol, currAngle), action1, 1.12},
                 {ActionMove60Up(currRow, currCol, currAngle), action2, 1.12},
                 {ActionMove30Down(currRow, currCol, currAngle), action3, 1.12},
                 {ActionMove60Down(currRow, currCol, currAngle), action4, 1.12},
                 {ActionMoveStraight(currRow, currCol, currAngle), currAngle, 1.0}})
        {
            if (action_valid)
            {
                double nextRow = currRow + std::round(move_i);
                double nextCol = currCol + std::round(move_j);
                Node nextNode = {nextRow, nextCol, next_angle};
                double g_next = distMap[currNode] + step_cost;

                if (!visited[nextNode] && distMap[nextNode] > g_next)
                {
                    double h_next = std::sqrt(std::pow(nextRow - goal.first, 2) +
                                              std::pow(nextCol - goal.second, 2));
                    distMap[nextNode] = g_next;
                    pathMap[nextNode] = currNode;
                    pq.push({g_next + h_next, nextNode});
                }
            }
        }

        std::vector<std::tuple<bool, int, double>> actions = {
            {ActionMove30Up(currRow, currCol, currAngle), action1, 1.12},
            {ActionMove60Up(currRow, currCol, currAngle), action2, 1.12},
            {ActionMove30Down(currRow, currCol, currAngle), action3, 1.12},
            {ActionMove60Down(currRow, currCol, currAngle), action4, 1.12},
            {ActionMoveStraight(currRow, currCol, currAngle), currAngle, 1.0}};
    }

    // No-path check: find the angle at goalRow/goalCol with the lowest g-cost
    int bestAngle = -1;
    double bestDist = std::numeric_limits<double>::infinity();
    for (int angle = 0; angle < 360; angle += 30)
    {
        Node n = {goalRow, goalCol, angle};
        auto it = distMap.find(n);
        if (it != distMap.end() && it->second < bestDist)
        {
            bestDist = it->second;
            bestAngle = angle;
        }
    }

    if (bestAngle == -1)
    {
        std::cout << "No path found\n";
        return {explored_states, {}, std::numeric_limits<double>::infinity()};
    }

    // Backtrack from goal to start through pathMap, then reverse
    std::vector<Node> backtrack_states;
    Node curr = {goalRow, goalCol, bestAngle};
    while (curr != SENTINEL)
    {
        backtrack_states.push_back(curr);
        curr = pathMap[curr];
    }
    std::reverse(backtrack_states.begin(), backtrack_states.end());

    return {explored_states, backtrack_states, bestDist};
}

// ---------------------------------------------------------------------------
// Animate
// ---------------------------------------------------------------------------
void Astar::Animate(const std::vector<Node> &explored,
                    const std::vector<Node> &backtrack,
                    const std::string &videoPath,
                    const std::string &imagePath)
{
    int fourcc = cv::VideoWriter::fourcc('X', 'V', 'I', 'D');
    cv::VideoWriter out(videoPath, fourcc, 20.0, cv::Size(numCols, numRows));
    cv::Mat image = cv::Mat::zeros(numRows, numCols, CV_8UC3);

    // Draw explored states in orange (BGR: 255, 150, 0)
    int count = 0;
    for (const auto &state : explored)
    {
        int r = numRows - (int)std::get<0>(state);
        int c = (int)std::get<1>(state) - 1;
        image.at<cv::Vec3b>(r, c) = {255, 150, 0};
        if (count % 75 == 0)
            out.write(image);
        count++;
    }

    // Draw free (unvisited, non-obstacle) cells in green (BGR: 154, 250, 0)
    for (int row = 1; row <= numRows; row++)
    {
        for (int col = 1; col <= numCols; col++)
        {
            if (image.at<cv::Vec3b>(numRows - row, col - 1) == cv::Vec3b{0, 0, 0} && IsValid(row, col) && !IsObstacle(row, col))
            {
                image.at<cv::Vec3b>(numRows - row, col - 1) = {154, 250, 0};
                if (count % 75 == 0)
                    out.write(image);
                count++;
            }
        }
    }

    // Draw backtrack path in blue (BGR: 0, 0, 255)
    for (const auto &state : backtrack)
    {
        int r = numRows - (int)std::get<0>(state);
        int c = (int)std::get<1>(state) - 1;
        image.at<cv::Vec3b>(r, c) = {0, 0, 255};
        out.write(image);
        cv::imshow("result", image);
        cv::waitKey(5);
    }

    if (!imagePath.empty())
        cv::imwrite(imagePath, image);

    cv::waitKey(0);
    cv::destroyAllWindows();
    out.release();
}

File: Python/astar_class.py
# header files
import cv2
import numpy as np 
import math
from heapq import heappush, heappop



# class for Astar
class Astar(object):
    # init function
    def __init__(self, start, goal, clearance, radius):
        self.start = start
        self.goal = goal
        self.numRows = 200
        self.numCols = 300
        self.i = 0
        self.j = 0
        self.clearance = clearance
        self.radius = radius
       
        
    # move is valid 
    def IsValid(self, currRow, currCol):
        return (currRow >= (1 + self.radius + self.clearance) and currRow <= (self.numRows - self.radius - self.clearance) and currCol >= (1 + self.radius + self.clearance) and currCol <= (self.numCols - self.radius - self.clearance))

    # checks for an obstacle
    def IsObstacle(self, row, col):
        # constants
        sum_of_c_and_r = self.clearance + self.radius
        sqrt_of_c_and_r = 1.4142 * sum_of_c_and_r
        
        # check circle
        dist1 = ((row - 150) * (row - 150) + (col - 225) * (col - 225)) - ((25 + sum_of_c_and_r) * (25 + sum_of_c_and_r))
        
        # check eclipse
        dist2 = ((((row - 100) * (row - 100)) / ((20 + sum_of_c_and_r) * (20 + sum_of_c_and_r))) + (((col - 150) * (col - 150)) / ((40 + sum_of_c_and_r) * (40 + sum_of_c_and_r)))) - 1
        
        # check triangles
        (x1, y1) = (120 - (2.62 * sum_of_c_and_r), 20 - (1.205 * sum_of_c_and_r))
        (x2, y2) = (150 - sqrt_of_c_and_r, 50)
        (x3, y3) = (185 + sum_of_c_and_r, 25 - (sum_of_c_and_r * 0.9247))
        first = ((col - y1) * (x2 - x1)) - ((y2 - y1) * (row - x1))
        second = ((col - y2) * (x3 - x2)) - ((y3 - y2) * (row - x2))
        third = ((col - y3) * (x1 - x3)) - ((y1 - y3) * (row - x3))
        dist3 = 1
        if(first <= 0 and second <= 0 and third <= 0):
            dist3 = 0
            
        (x1, y1) = (150 - sqrt_of_c_and_r, 50)
        (x2, y2) = (185 + sum_of_c_and_r, 25 - (sum_of_c_and_r * 0.9247))
        (x3, y3) = (185 + sum_of_c_and_r, 75 + (sum_of_c_and_r * 0.5148))
        first = ((col - y1) * (x2 - x1)) - ((y2 - y1) * (row - x1))
        second = ((col - y2) * (x3 - x2)) - ((y3 - y2) * (row - x2))
        third = ((col - y3) * (x1 - x3)) - ((y1 - y3) * (row - x3))
        dist4 = 1
        if(first >= 0 and second >= 0 and third >= 0):
            dist4 = 0
        
        # check rhombus
        (x1, y1) = (10 - sqrt_of_c_and_r, 225)
        (x2, y2) = (25, 200 - sqrt_of_c_and_r)
        (x3, y3) = (40 + sqrt_of_c_and_r, 225)
        (x4, y4) = (25, 250 + sqrt_of_c_and_r)
        first = ((col - y1) * (x2 - x1)) - ((y2 - y1) * (row - x1))
        second = ((col - y2) * (x3 - x2)) - ((y3 - y2) * (row - x2))
        third = ((col - y3) * (x4 - x3)) - ((y4 - y3) * (row - x3))
        fourth = ((col - y4) * (x1 - x4)) - ((y1 - y4) * (row - x4))
        dist5 = 1
        dist6 = 1
        if(first >= 0 and second >= 0 and third >= 0 and fourth >= 0):
            dist5 = 0
            dist6 = 0
        
        # check square
        (x1, y1) = (150 - sqrt_of_c_and_r, 50)
        (x2, y2) = (120 - sqrt_of_c_and_r, 75)
        (x3, y3) = (150, 100 + sqrt_of_c_and_r)
        (x4, y4) = (185 + sum_of_c_and_r, 75 + (sum_of_c_and_r * 0.5148))
        first = ((col - y1) * (x2 - x1)) - ((y2 - y1) * (row - x1))
        second = ((col - y2) * (x3 - x2)) - ((y3 - y2) * (row - x2))
        third = ((col - y3) * (x4 - x3)) - ((y4 - y3) * (row - x3))
        fourth = ((col - y4) * (x1 - x4)) - ((y1 - y4) * (row - x4))
        dist7 = 1
        dist8 = 1
        if(first <= 0 and second <= 0 and third <= 0 and fourth <= 0):
            dist7 = 0
            dist8 = 0
        
        # check rod
        first = ((col - 95) * (8.66 + sqrt_of_c_and_r)) - ((5 + sqrt_of_c_and_r) * (row - 30 + sqrt_of_c_and_r))
        second = ((col - 95) * (37.5 + sqrt_of_c_and_r)) - ((-64.95 - sqrt_of_c_and_r) * (row - 30 + sqrt_of_c_and_r))
        third = ((col - 30.05 + sqrt_of_c_and_r) * (8.65 + sqrt_of_c_and_r)) - ((5.45 + sqrt_of_c_and_r) * (row - 67.5))
        fourth = ((col - 35.5) * (-37.49 - sqrt_of_c_and_r)) - ((64.5 + sqrt_of_c_and_r) * (row - 76.15 - sqrt_of_c_and_r))
        dist9 = 1
        dist10 = 1
        if(first <= 0 and second >= 0 and third >= 0 and fourth >= 0):
            dist9 = 0
            dist10 = 0
        
        if(dist1 <= 0 or dist2 <= 0 or dist3 == 0 or dist4 == 0 or dist5 == 0 or dist6 == 0 or dist7 == 0 or dist8 == 0 or dist9 == 0 or dist10 == 0):
            return True
        return False

    # action move left
    def ActionMove30Up(self, currRow, currCol, currangle):
        theta = 30 + currangle
        if(theta >= 360):
            theta = theta - 360
        rad = theta * (math.pi/180)
        self.i = math.sin(rad)
        self.j = math.cos(rad)

        if(self.IsValid(currRow + self.i, currCol + self.j) and self.IsObstacle(currRow + self.i, currCol + self.j) == False):
            return True
        return False

    def ActionMove60Up(self, currRow,currCol, currangle):
        theta = 60 + currangle
        if(theta >= 360):
            theta = theta - 360
        rad = theta * (math.pi/180)
        self.i = math.sin(rad)
        self.j = math.cos(rad)

        if(self.IsValid(currRow + self.i, currCol + self.j) and self.IsObstacle(currRow + self.i, currCol + self.j) == False):
            return True
        return False

    def ActionMove30Down(self, currRow,currCol, currangle):
        theta = currangle - 30
        if(theta < 0):
            theta = 360 + theta
        rad = theta * (math.pi/180)
        self.i = math.sin(rad)
        self.j = math.cos(rad)

        if(self.IsValid(currRow + self.i, currCol + self.j) and self.IsObstacle(currRow + self.i, currCol + self.j) == False):
            return True
        return False

    def ActionMove60Down(self, currRow,currCol, currangle):
        theta = currangle - 60
        if(theta < 0):
            theta = 360 + theta
        rad = theta * (math.pi/180)
        self.i = math.sin(rad)
        self.j = math.cos(rad)

        if(self.IsValid(currRow + self.i, currCol + self.j) and self.IsObstacle(currRow + self.i, currCol + self.j) == False):
            return True
        return False

    def ActionMoveStraight(self, currRow,currCol, currangle):
        theta = currangle
        rad = theta * (math.pi/180)
        self.i = math.sin(rad)
        self.j = math.cos(rad)
        if(self.IsValid(currRow + self.i, currCol + self.j) and self.IsObstacle(currRow + self.i , currCol + self.j) == False):
            return True
        return False


    def CheckIfGoal(self, currRow, currCol):
        check = (((currRow - self.goal[0]) * (currRow - self.goal[0])) + ((currCol - self.goal[1]) * (currCol - self.goal[1])) - ( 1.5 * 1.5))
        if(check <= 0):
            global goalRow
            global goalCol
            goalRow = currRow
            goalCol = currCol
            print("goal reached")
            return True
        else:
            return False


    
    # astar algorithm
    def Astar(self):
        # create hashmap to store distances
        distMap = {}
        visited = {}
        path = {}
        for row in np.arange(1, self.numRows + 1, 0.5):
            for col in np.arange(1, self.numCols + 1, 0.5):
                for angle in range(0,360,30):
                    distMap[(row, col, angle)] = float('inf')
                    path[(row, col, angle)] = -1
                    visited[(row, col, angle)] = False
            
        # create queue, push the source and mark distance from source to source as zero
        explored_states = []
        queue = []
        heappush(queue, (0, self.start))
        distMap[self.start] = 0

        while(len(queue) > 0):
            _, currNode = heappop(queue)
            if visited[currNode]:
                continue
            visited[currNode] = True
            explored_states.append(currNode)

            # if goal node then exit
            if(self.CheckIfGoal(currNode[0],currNode[1]) == True):
                break

            # go through each edge of current node
            a1 = (currNode[2] + 30) % 360
            a2 = (currNode[2] + 60) % 360
            a3 = (currNode[2] - 30) % 360
            a4 = (currNode[2] - 60) % 360

            moves = [
                (self.ActionMove30Up,    a1, 1.12),
                (self.ActionMove60Up,    a2, 1.12),
                (self.ActionMove30Down,  a3, 1.12),
                (self.ActionMove60Down,  a4, 1.12),
                (self.ActionMoveStraight, currNode[2], 1.0),
            ]

            for action, next_angle, step_cost in moves:
                if action(currNode[0], currNode[1], currNode[2]):
                    nextRow = currNode[0] + round(self.i)
                    nextCol = currNode[1] + round(self.j)
                    nextNode = (nextRow, nextCol, next_angle)
                    g_next = distMap[currNode] + step_cost
                    if not visited[nextNode] and distMap[nextNode] > g_next:
                        h_next = math.sqrt((nextRow - self.goal[0])**2 + (nextCol - self.goal[1])**2)
                        distMap[nextNode] = g_next
                        path[nextNode] = currNode
                        heappush(queue, (g_next + h_next, nextNode))




        # return if no optimal path
        check = []
        f1 = self.goal[0]
        f2 = self.goal[1]
        for a in np.arange(f1-1,f1+1.5,0.5):
            for b in np.arange(f2-1,f2+1.5,0.5):
                for c in range(0,360,30):
                    check.append(distMap[a,b,c])
        for c in range(0,360,30):
            check.append(distMap[f1,f2-1.5,c])
        for c in range(0,360,30):
            check.append(distMap[f1,f2+1.5,c])
        for c in range(0,360,30):
            check.append(distMap[f1+1.5,f2,c])
        for c in range(0,360,30):
            check.append(distMap[f1-1.5,f2,c])
        NoPath = 0
        ans = float('inf')
        for a in range(len(check)):
            if(check[a] != ans):
                print("There exists a path")
                NoPath = 1
                break
        if(NoPath == 0):
            print("NO VALID PATH")
            return (explored_states, [], distMap[f1,f2,0])

        for a in range(0,360,30):
            if(distMap[goalRow,goalCol,a] != ans):
                bird = a

        print(distMap[goalRow,goalCol,bird],"answer")
        result = (goalRow, goalCol, bird)
        
        # backtrack path
        backtrack_states = []
        node = result
        while(path[node] != -1):
            backtrack_states.append(node)
            node = path[node]
        backtrack_states.append(self.start)
        backtrack_states = list(reversed(backtrack_states))
        # print(explored_states) 
        return (explored_states, backtrack_states, distMap[goalRow,goalCol,bird])
    

    # animate path
    def animate(self, explored_states, backtrack_states, path, image_path=None):
        path = str(path)

        # Create blank frame (height, width, channels)
        image = np.zeros((self.numRows, self.numCols, 3), dtype=np.uint8)
        fourcc = cv2.VideoWriter_fourcc(*'mp4v')
        out = cv2.VideoWriter(str(path), fourcc, 20.0, (self.numCols,self.numRows))
        image = np.zeros((self.numRows, self.numCols, 3), dtype=np.uint8)


        # If MP4 failed (missing codecs), fall back to AVI
        if not out.isOpened():
            if not path.lower().endswith('.avi'):
                path = path.rsplit('.', 1)[0] + '.avi'
            fourcc = cv2.VideoWriter_fourcc(*'XVID')
            out = cv2.VideoWriter(path, fourcc, fps, (self.numCols, self.numRows))

        if not out.isOpened():
            raise RuntimeError("cv2.VideoWriter failed to open (MP4 and AVI). Install ffmpeg/gstreamer codecs or use XVID/MJPG.")

        count = 0
        for state in explored_states:
            image[int(self.numRows - state[0]), int(state[1] - 1)] = (255, 150, 0)
            if(count%75 == 0):
                out.write(image)
            count = count + 1

        count = 0
        for row in range(1, self.numRows + 1):
            for col in range(1, self.numCols + 1):
                if(image[int(self.numRows - row), int(col - 1), 0] == 0 and image[int(self.numRows - row), int(col - 1), 1] == 0 and image[int(self.numRows - row), int(col - 1), 2] == 0):
                    if(self.IsValid(row, col) and self.IsObstacle(row, col) == False):
                        image[int(self.numRows - row), int(col - 1)] = (154, 250, 0)
                        if(count%75 == 0):
                            out.write(image)
                        count = count + 1
            
        if(len(backtrack_states) > 0):
            for state in backtrack_states:
                image[int(self.numRows - state[0]), int(state[1] - 1)] = (0, 0, 255)
                out.write(image)
                cv2.imshow('result', image)
                cv2.waitKey(5)
                
        if image_path is not None:
            cv2.imwrite(str(image_path), image)
        cv2.waitKey(0)
        cv2.destroyAllWindows()
        out.release()

        print("Saved video to:", path)
